shared/subspace-data-retrieval/src/object_fetcher/tests.rs

```diff
@@ -0,0 +1,829 @@
+
+    let mut bytes_written = 0;
+
+    if let Some((prefix_bytes, padding_bytes)) = skip_padding {
+        let mut raw_data = extract_raw_data_mut(pieces);
+
+        (&mut raw_data)
+            .skip(offset)
+            .take(prefix_bytes)
+            .zip(object_len_encoded.iter().take(prefix_bytes))
+            .for_each(|(raw_data_byte, len_byte)| {
+                *raw_data_byte = *len_byte;
+                bytes_written += 1;
+            });
+
+        assert_eq!(bytes_written, prefix_bytes);
+
+        raw_data
+            .skip(padding_bytes)
+            .take(object_len_encoded_len - prefix_bytes)
+            .zip(object_len_encoded.iter().skip(prefix_bytes))
+            .for_each(|(raw_data_byte, len_byte)| {
+                *raw_data_byte = *len_byte;
+                bytes_written += 1;
+            });
+
+        assert_eq!(bytes_written, object_len_encoded_len);
+    } else {
+        let raw_data = extract_raw_data_mut(pieces);
+        raw_data
+            .skip(offset)
+            .take(object_len_encoded_len)
+            .zip(object_len_encoded)
+            .for_each(|(raw_data_byte, len_byte)| {
+                *raw_data_byte = len_byte;
+                bytes_written += 1;
+            });
+
+        assert_eq!(bytes_written, object_len_encoded_len);
+    }
+}
+
+/// Creates a mapping from piece(s), start piece index, offset, and object length.
+/// Returns the mapping and object data.
+///
+/// If supplied:
+/// - `skip_padding` is the amount of padding to skip, and
+/// - `raw_data_after_segment_header` is the raw data from the piece at the start of the segment,
+///   after the segment header has been written to it.
+fn create_mapping(
+    pieces: Vec<&Piece>,
+    start_piece_index: usize,
+    offset: usize,
+    object_len: usize,
+    mut skip_padding: Option<usize>,
+    raw_data_after_segment_header: Option<Vec<u8>>,
+) -> (GlobalObject, Vec<u8>) {
+    // Simplify the code below by setting the default skip padding value.
+    if raw_data_after_segment_header.is_some() && skip_padding.is_none() {
+        skip_padding = Some(0);
+    }
+    if skip_padding.is_some() {
+        assert!(
+            raw_data_after_segment_header.is_some(),
+            "can only skip padding if there is more data in the next segment"
+        );
+    }
+
+    let object_len_encoded = compact_encoded(object_len);
+    let pieces_len = pieces.len();
+    let mut raw_data = extract_raw_data(pieces).collect::<Vec<u8>>();
+
+    if let (Some(skip_padding), Some(raw_data_after_segment_header)) =
+        (skip_padding, raw_data_after_segment_header)
+    {
+        // Find the next piece index with a segment header
+        let segment_header_piece_index =
+            start_piece_index.next_multiple_of(ArchivedHistorySegment::NUM_PIECES);
+        // And the piece before it can have padding
+        let padding_piece_index = segment_header_piece_index.checked_sub(1);
```

Comment:
This doesn't really make sense to me. Subtracting 1 will hit a parity piece index (the last piece in the previous segment), it doesn't have source data at all.

Reply:
This should be 2, I'll fix it today.